### PR TITLE
feat(core/meetings): surfaces — CLI/MCP/HTTP + adapter + tail-step (#1900)

### PR DESCRIPTION
**Pull request by joshuaswarren**
Surfaces slice of the meeting-intelligence feature (issue #1900): CLI/MCP/HTTP surfaces + the activity/wearables day-source adapter + the post-sync build tail-step, plus memory generation (deterministic episode + trust-gated summary/facts) wired onto the engine seam.

Stacked on the deterministic-engine branch feat/meetings-engine (merge that first). Kept as a cross-cutting registration PR by design; meeting surface registration inherently touches the shared CLI/MCP/HTTP dispatchers, extracted into meeting sibling modules so the dispatchers stay within their structural size ceilings.

Verified: build + typecheck clean; scoped ratchet OK (baseline untouched, oversized-file count unchanged); config-contract OK; focused suites green (398). Part of issue #1900.

---

> [!NOTE]
> **High Risk**
> Cross-cutting access surfaces, multi-tenant namespace isolation, mutating meeting builds (records + memories + reindex), and LLM-backed summary/facts increase blast radius if scope or flush/debounce behavior regresses.
> 
> **Overview**
> Exposes **retrospective meetings** (`list` / `show` / `build`) on the standalone `remnic` CLI, host `engram` CLI, HTTP (`/engram/v1/meetings*`), and MCP (`engram.meetings_*`), with shared runners in `@remnic/core` so behavior stays aligned across entrypoints.
> 
> **Namespace-scoped wearables + meetings (#2123):** HTTP/MCP/operations now pass caller `namespace` / `sessionKey` / principal into wearables and meetings APIs so reads and writes land in the caller’s namespace (not a machine-wide default). Non-default callers do not fuse default-namespace wearables or global activity.
> 
> **Post-sync meeting builds:** Wearable and activity sync schedule debounced per-day rebuilds via `MeetingsBuildScheduler`. Activity sync reports `touchedDates` and optional `onSynced` so backfilled older days rebuild too. **Manual** one-shot `wearables sync` (CLI/HTTP/`EngramAccessService`) calls `flushBuilds()` when meetings are enabled so short-lived processes don’t exit before the unref’d debounce timer; long-lived auto-sync keeps debouncing.
> 
> Adds the **activity + wearable day-source adapter** (`ActivityWearablesMeetingsDaySource`) and **memory generation** (deterministic episode memories plus trust-gated summary/facts on the builder seam). HTTP `meetings/build` uses the write rate limit; malformed meeting IDs return 400.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3e54c84a37e5d0983dfcc26d717557a982affcf6. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added meetings management through the CLI, HTTP API, and MCP tools, including list, view, and build operations.
  * Added automatic meeting detection and memory generation from wearable transcripts and activity data.
  * Added configurable meeting summaries with trust-gated facts and review workflows.
  * Added namespace-aware meeting and transcript access for improved data isolation.

* **Bug Fixes**
  * Manual wearable sync now flushes pending meeting builds before completion.
  * Improved validation and error responses for malformed meeting requests.
  * Prevented cross-namespace transcript and meeting results.

### CODE REVIEW

File: packages/remnic-core/src/orchestration/workspace-ops.ts [hunk 1]
```diff
@@ -688,6 +704,82 @@ export class WorkspaceOpsCoordinator {
...
+            // Meeting records live in the QMD collection root; reindex on change
+            // so `show`/search find them immediately (qmd.update clears its own
+            // result caches). Fired only when records changed; a failure is
+            // surfaced as a build warning, not a hard failure.
+            await this.deps.qmd.update();
+          },
+        },
+        // The same sealed write path the wearables generator uses, so meeting
+        // episode memories inherit encrypted-at-rest + atomic write + dedup.
+        memoryWriter: storage,
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/workspace-ops.ts]: Fixed in 6139856b2: production episode idempotency holds end-to-end (createMeetingMemoryGenerator + createMeetingMemoryWriter dedupe by source(meeting:<id>)+content) AND the unchangedIds seam skips all generation on an unchanged-day rebuild. Regression: build-memory.test.ts 'finding 1 — two production builds of an unchanged day write the episode once' + 'finding 2 — an unchanged-record rebuild fires zero LLM calls and writes nothing'.

File: packages/remnic-cli/src/commands/meetings.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      meetingsService = await meetingsOrchestrator.getMeetingsService();
+    } catch {
+      console.error(
+        "meetings: failed to load the Remnic config or start the memory engine — run `remnic doctor` and check the config file for errors",
+      );
+      process.exitCode = 1;
+      return;
+    }
+    const code = await runMeetingsCliCommand(
+      { store: meetingsService.store, builder: meetingsService.builder },
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-cli/src/commands/meetings.ts]: Fixed: the standalone binary CLI now routes through the shared runner which gates list/show on deps.config.enabled (MeetingsCliDeps.config is required); disabled => no records exposed. Regression: meetings/cli.test.ts enabled-gate cases.

File: packages/remnic-core/src/orchestrator.ts [hunk 3]
```diff
@@ -1486,6 +1489,7 @@ export class Orchestrator {
       namespaceSearchRouter: this.namespaceSearchRouter,
       namespaceCatalog: this.namespaceCatalog,
       getStorage: () => this.storage, storageForNamespace: (namespace) => this.storageRouter.storageFor(namespace),
+      onActivitySynced: (summary) => { if (!this.config.meetings.enabled) return; void this.getMeetingsService().then((s) => s.requestBuildForActivitySync(summary.ranAt, this.config.activity.timezone)).catch((err) => log.warn(`meetings: activity tail-step build failed to schedule: ${err instanceof Error ? err.message : String(err)}`)); },
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestrator.ts]: Fixed in 6139856b2: the activity runner now reports touchedDates and MeetingsService.requestBuildForActivitySync fans out a build over every touched date (not just ranAt's day); orchestrator forwards summary.touchedDates. Regression: activity/runner.test.ts 'runActivitySyncOnce reports every local date that gained snapshots' + service.test.ts 'requestBuildForActivitySync rebuilds every touched date'.

File: packages/remnic-core/src/access-http.ts [hunk 4]
```diff
@@ -1610,6 +1600,15 @@ export class EngramAccessHttpServer {
       return;
     }
 
+    if (req.method === "GET" && (pathname === "/engram/v1/meetings" || pathname === "/remnic/v1/meetings")) {
+      this.enforceTokenOp("meetings_list"); await respondMeetingsList(res, this.respondJson.bind(this), this.service, nonEmptyQueryParam(parsed.searchParams.get("date"))); return; }
+    if (req.method === "POST" && (pathname === "/engram/v1/meetings/build" || pathname === "/remnic/v1/meetings/build")) {
+      this.enforceTokenOp("meetings_build"); await respondMeetingsBuild(req, res, this.respondJson.bind(this), this.readJsonBody.bind(this), this.service); return; }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/access-http.ts]: Fixed in 6139856b2: POST /engram|remnic/v1/meetings/build now enforces the per-principal write quota (ensureWriteRateLimitAvailable before, recordWriteRateLimitHit after) via the http-glue write-quota hooks, matching other mutating REST routes. Regression: access-http.test.ts 'HTTP meetings build enforces the per-principal write rate limit'.

File: packages/remnic-core/src/meetings/day-source.ts [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+      }
+    }
+    const endUtc =
+      conversation.endIso ??
+      (Number.isFinite(maxEndMs) ? new Date(maxEndMs).toISOString() : conversation.startIso);
+    return {
+      source: conversation.source,
+      startUtc: conversation.startIso,
+      endUtc,
+      distinctNonWearerSpeakers: speakers.size,
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/meetings/day-source.ts]: Fixed in f7ff3fc02: buildAudioWindows now carries conversation.title and marks known cloud-provider sources (granola/fireflies) with providerMeeting:true, so a titled provider conversation with no app span fires detect's provider branch. Regression: day-source.test.ts 'buildAudioWindows marks provider sources + carries the title so a titled provider conversation with no app span is detected'.

File: packages/remnic-core/src/orchestration/orchestrator-init.ts [hunk 6]
```diff
@@ -562,7 +564,14 @@ export class OrchestratorInitCoordinator {
               : {}),
           },
           {
-            sync: (options) => this.deps.getWearablesService().sync(options),
+            sync: async (options) => {
+              const summaries = await this.deps.getWearablesService().sync(options);
+              // Tail step (issue #1900): a wearable sync changed a day's audio,
+              // so schedule a debounced meeting rebuild for each synced day. The
+              // scheduler coalesces bursts; a failure here never fails the sync.
+              await this.scheduleMeetingBuildsForDays(summaries.flatMap((s) => s.days));
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/orchestrator-init.ts]: Fixed in f7ff3fc02: the meeting tail-step moved onto the shared WearablesService via an onDaysSynced hook fired in sync(), so BOTH auto-sync AND the manual HTTP/MCP/CLI wearables-sync path (getWearablesService().sync) rebuild affected meetings (gated on meetings.enabled). Regression: wearables/service.test.ts 'sync fires onDaysSynced with the synced days so a MANUAL wearables sync triggers the meeting tail-step'.

File: packages/remnic-core/src/orchestration/workspace-ops.ts [hunk 7]
```diff
@@ -688,6 +706,109 @@ export class WorkspaceOpsCoordinator {
...
+   * — CLI, MCP tools, HTTP routes — shares this one instance. Writes are pinned
+   * to the same deterministic namespace bulk-import uses, so meeting records and
+   * episode memories land beside wearable/activity artifacts. Constructing it
+   * touches no disk while `meetings.enabled` is off — every entrypoint gates.
+   */
+  async getMeetingsService(): Promise<MeetingsService> {
+    if (!this.deps.meetingsServiceInstance) {
+      const config = this.deps.config.meetings;
+      const memoryDir = this.deps.config.memoryDir;
+      const storage = await this.deps.getStorageForNamespace(this.deps.bulkImportWriteNamespace());
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/workspace-ops.ts]: Fixed in c08c3a7c6 (#2123): caller-derived namespace symmetry implemented. getWearablesService/getMeetingsService now take a namespace and cache per-namespace; storage/records/wearable-source/episode+summary memories are scoped to the CALLER namespace (resolveReadableNamespace for reads, writableNamespaceFor for writes, threaded from the HTTP/MCP principal). Screen activity stays machine-scoped and is consumed ONLY by the default/machine-owner namespace; non-default callers are strictly isolated (no default-ns wearable fallback, no global activity). Documented in docs/meetings.md. Regressions: meetings/namespace-isolation.test.ts (two-principal physical isolation: A's data under A's root, B sees none, non-default caller gets neither default wearables nor global activity, machine-owner does) + testing/subjects/meetings-namespace.test.ts (runLifecycleMatrix).

File: packages/remnic-core/src/meetings/memory-gen.ts [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+      // Retract the prior memories of records that changed under a kept id, so the
+      // regeneration below REFRESHES the meeting (never leaves stale episodes/facts
+      // beside the new ones). The updated records are still in `built` and regenerate.
+      for (const updatedId of updatedIds) {
+        await writer.retireMemoriesFromSource(meetingSourceLabel(updatedId));
+      }
+      // Skip unchanged records entirely: an idempotent rebuild must fire no LLM and
+      // rewrite nothing. Generation runs only on changed (updated) + new records.
+      const unchanged = new Set(unchangedIds);
+      const toGenerate = built.filter((record) => !unchanged.has(record.id));
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/meetings/memory-gen.ts]: Fixed in c08c3a7c6 (#2123): unchanged ids now skip generation ONLY when their memories already exist (probed via MeetingMemoryWriter.hasMemoryFromSource on the deterministic episode). A build that persisted the record but threw before memories landed is regenerated on the next build. Regression: memory-gen.test.ts 'finding 2b — an unchanged id whose memories are MISSING is regenerated (retryable after a prior generation throw)'.

**Comment by joshuaswarren** [comment]
scope-budget-exempt justification: Caller-namespace symmetry P1 was required in the existing surfaces PR after PR proliferation was frozen; this cross-cutting access/storage wiring is not separable without another PR.

File: packages/remnic-core/src/orchestration/workspace-ops.ts [hunk 9]
```diff
@@ -683,11 +701,134 @@ export class WorkspaceOpsCoordinator {
...
+        // BOTH the auto-sync adapter and the manual HTTP/MCP/CLI sync path
+        // (EngramAccessService.wearablesSync -> getWearablesService().sync)
+        // rebuild affected meetings. Gated on meetings.enabled so a sync of
+        // other signals never spins up meeting building; a failure never fails
+        // the sync.
+        onDaysSynced: async (days) => {
+          if (!this.deps.config.meetings.enabled) return;
+          try {
+            const meetings = await this.getMeetingsService();
+            for (const day of days) meetings.requestBuild(day);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/workspace-ops.ts:732]: Fixed in 24fc5dfdf (#2123): the manual one-shot wearables-sync path now DRAINS the debounced meeting tail-step build before the short-lived caller exits — access-service.wearablesSync (shared HTTP/MCP manual completion) and the standalone cli.ts forwardWearables 'sync' path both call MeetingsService.flushBuilds() (gated on meetings.enabled) after sync; the long-lived auto-sync daemon path keeps debouncing (untouched). Regressions: meetings/manual-sync-flush.test.ts 'manual wearables sync flushes the debounced meeting build' + 'auto-sync path keeps the meeting build debounced'.

File: packages/remnic-core/src/meetings/service.ts [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    this.scheduler = new MeetingsBuildScheduler({
+      debounceMs: deps.buildDebounceMs,
+      build: (date) => this.builder.buildDay(date),
+      ...(deps.onBuildError ? { onError: deps.onBuildError } : {}),
+    });
+  }
+
+  /** List stored meetings for one day, or across all days when no date given. */
+  async meetingsList(date?: string): Promise<MeetingsListResult> {
+    if (!this.config.enabled) return { enabled: false, days: [] };
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/meetings/service.ts:77]: Acknowledged (P2, validation-consistency). Per maintainer hard-stop on further code for lower-severity findings, this is accepted for now and captured as a follow-up: meetingsList should validate the --date/date input before the meetings.enabled=false empty-success short-circuit (the enabled path already rejects a malformed date; the CLI runner was aligned in an earlier round, this is the service-layer parity). No data isolation/loss; core contract + regressions unaffected.

File: packages/remnic-core/src/access-mcp.ts [hunk 11]
```diff
@@ -184,7 +184,7 @@ const MCP_MIGRATED_OPERATIONS: Readonly<Record<string, OperationName>> = {
   "engram.wearables_sync": "wearables_sync",
   "engram.transcript_day": "transcript_day",
   "engram.transcript_search": "transcript_search",
-  "engram.transcript_memories": "transcript_memories",
+  "engram.transcript_memories": "transcript_memories", "engram.meetings_list": "meetings_list", "engram.meetings_get": "meetings_get", "engram.meetings_build": "meetings_build",
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/access-mcp.ts:187]: Acknowledged (P2, quota-billing parity). Per maintainer hard-stop, captured as a follow-up: add engram.meetings_build/remnic.meetings_build to the MCP write predicate (isMcpWrite) so the MCP build path is billed against the per-principal write quota, matching the HTTP build route (which already enforces the quota). This is quota-accounting parity, not a data-isolation/loss defect; no test regression.

File: packages/remnic-core/src/meetings/service.ts [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    const days: MeetingsListDay[] = [];
+    for (const day of dates) {
+      days.push({ date: day, meetings: await this.store.listMeetingSummaries(day) });
+    }
+    return { enabled: true, days };
+  }
+
+  /** Fetch one stored meeting record's raw markdown by id. */
+  async meetingsGet(id: string): Promise<MeetingsGetResult> {
+    if (!this.config.enabled) return { enabled: false, found: false, id, record: null };
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/meetings/service.ts:94]: Acknowledged (P2, validation-consistency, same class as the earlier disabled-short-circuit findings). Per maintainer hard-stop on lower-severity code changes, captured as a follow-up: meetingsGet (and meetingsList) should validate the id/date shape before the meetings.enabled=false disabled/absent short-circuit, so a malformed id/date is rejected regardless of enabled state (the enabled path + the CLI runner already reject it). Not a data-isolation/loss or test-failure; core contract + regressions unaffected.

File: packages/remnic-core/src/meetings/memory-gen.ts [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+          memory.frontmatter.status !== "rejected" &&
+          stripAttributesSuffix(memory.content) === needle,
+      );
+    },
+    retireMemoriesFromSource: async (source: string): Promise<number> => {
+      const memories = await storage.readAllMemories();
+      let retired = 0;
+      for (const memory of memories) {
+        if (memory.frontmatter.source !== source) continue;
+        if (await storage.invalidateMemory(memory.frontmatter.id)) retired++;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/meetings/memory-gen.ts:422]: Acknowledged (P2, cleanup-state observability). Per maintainer hard-stop, captured as a follow-up: when invalidateMemory() returns false during removed/updated-meeting cleanup, surface the failed retirement (warn / re-queue) instead of silently treating the source as handled, so a failed retire is retried rather than leaving an orphaned meeting memory. Not data-isolation/loss or a test failure; retirement is best-effort today and the deterministic episode remains the presence sentinel for the retry path.

File: packages/remnic-cli/src/index.ts [hunk 14]
```diff
@@ -13330,6 +13332,11 @@ Other:
       break;
     }
 
+    case "meetings": {
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-cli/src/index.ts:13340]: Fixed in db04cbdef (#2123): the standalone `remnic wearables sync` binary path (main() case 'wearables') exited before the debounced meeting build fired. It now drains getMeetingsService().flushBuilds() after a one-shot sync (gated on meetings.enabled) before shutdown — mirroring the core cli.ts forwardWearables + access-service.wearablesSync paths; the long-lived auto-sync path stays debounced. index.ts held at its 13736 ceiling. Regressions: remnic-cli wearables-sync-flush.test.ts 'standalone remnic wearables sync flushes the debounced meeting build before the one-shot process exits' + the auto-sync-still-debounces guard.

File: packages/remnic-core/src/orchestration/workspace-ops.ts [hunk 15]
```diff
@@ -635,12 +653,12 @@ export class WorkspaceOpsCoordinator {
...
-    if (!this.deps.wearablesServiceInstance) {
-      this.deps.wearablesServiceInstance = new WearablesService({
+  getWearablesService(namespace: string): WearablesService {
+    let service = this.deps.wearablesServiceByNamespace.get(namespace);
+    if (!service) {
+      service = new WearablesService({
         config: this.deps.config.wearables,
-        getStorage: async () =>
-          await this.deps.getStorageForNamespace(this.deps.bulkImportWriteNamespace()),
+        getStorage: async () => await this.deps.getStorageForNamespace(namespace),
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/workspace-ops.ts:662]: Fixed in db04cbdef (#2123): the wearables transcript searchBackend was calling the ROOT qmd collection unconditionally, leaking root/other-namespace transcripts to non-default callers. It is now scoped to the caller namespace's collection (reuses namespaceCollectionName, mirrors qmdCollectionForHealth); namespaces-off and the default/machine-owner namespace resolve to the base collection byte-identically, and the search->null bounded-scan fallback is intact. Regression: wearables/transcript-search-namespace-isolation.test.ts 'transcript search is scoped to the caller namespace: A sees only A, B never sees the default transcript, the default caller still sees default' (proven failing pre-fix).

File: packages/remnic-core/src/orchestration/workspace-ops.ts [hunk 16]
```diff
@@ -683,9 +718,138 @@ export class WorkspaceOpsCoordinator {
...
+        source,
+        store,
+        config,
+        hooks: {
+          reindex: async () => {
+            // Meeting records live in the QMD collection root; reindex on change
+            // so `show`/search find them immediately (qmd.update clears its own
+            // result caches). Fired only when records changed; a failure is
+            // surfaced as a build warning, not a hard failure.
+            await this.deps.qmd.update();
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/workspace-ops.ts:826]: Acknowledged (P2, namespace reindex completeness). Per maintainer hard-stop on lower-severity code, captured as a follow-up: after a non-default-namespace meetingsBuild, the reindex hook refreshes only the default search backend; namespace-scoped backends (Orama/Meilisearch/LanceDB indexing per-namespace roots) should also be refreshed so caller-namespace meeting records are searchable immediately rather than at the next maintenance pass. Records ARE written correctly to the caller namespace (no data-loss/isolation issue) — this is search-freshness only.

File: packages/remnic-core/src/meetings/http-glue.ts [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    throw new EngramAccessInputError("date is required (YYYY-MM-DD)");
+  }
+  // Build persists/deletes records, writes memories, and reindexes — a mutating
+  // route, so it carries the same per-principal write quota as every other
+  // write (capsule import/export, admin promote). Enforce before the write,
+  // account the hit after it succeeds (issue #1937).
+  quota.enforceQuota();
+  const scope = scopeFor?.(
+    typeof body.namespace === "string" ? body.namespace : undefined,
+    typeof body.sessionKey === "string" ? body.sessionKey : undefined,
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/meetings/http-glue.ts:116]: Acknowledged (P2, input validation). Per maintainer hard-stop, captured as a follow-up: POST /meetings/build should reject (400) a non-string `namespace`/`sessionKey` rather than coercing to undefined (which falls back to the caller's default write namespace). No cross-namespace leak (it defaults to the caller's own scope, never another principal's), so not a data-isolation/loss defect; strict-input parity with the other routes is the follow-up.

File: packages/remnic-core/src/access-service.ts [hunk 18]
```diff
@@ -5729,60 +5743,62 @@ export class EngramAccessService {
...
+  async meetingsList(date?: string, scope?: WearablesMeetingsScope): Promise<MeetingsListResult> {
+    return (await this.orchestrator.getMeetingsService(this.resolveReadableNamespace(scope?.namespace, this.resolveRequestPrincipal(scope?.sessionKey, scope?.authenticatedPrincipal)))).meetingsList(date);
+  }
+
+  async meetingsGet(id: string, scope?: WearablesMeetingsScope): Promise<MeetingsGetResult> {
+    return (await this.orchestrator.getMeetingsService(this.resolveReadableNamespace(scope?.namespace, this.resolveRequestPrincipal(scope?.sessionKey, scope?.authenticatedPrincipal)))).meetingsGet(id);
+  }
+
+  async meetingsBuild(date: string, scope?: WearablesMeetingsScope): Promise<MeetingsDayBuildSummary> {
+    return (await this.orchestrator.getMeetingsService(this.writableNamespaceFor(scope?.namespace, scope?.sessionKey, scope?.authenticatedPrincipal))).meetingsBuild(date);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/access-service.ts]: Fixed in 3e54c84a3 (#2123): meetingsBuild now enforces the READ ACL on the resolved namespace (via the existing resolveReadableNamespace resolver) in ADDITION to writableNamespaceFor, since the build reads the namespace's stored wearable transcripts + existing meeting records. A principal that may write a namespace but is omitted from its readPrincipals is now DENIED meetingsBuild with the shared EngramAccessInputError ('namespace is not readable'). resolveReadableNamespace is a no-op ACL when namespaces are disabled, so default/operator behavior is unchanged. Regression: meetings/namespace-isolation.test.ts 'a write-only principal (in writePrincipals, not readPrincipals) is DENIED meetingsBuild, while read+write and the machine owner succeed' (proven failing pre-fix: stashing the ACL line -> 'Missing expected rejection', passing post-fix).

File: packages/remnic-core/src/access-mcp.ts [hunk 19]
```diff
@@ -687,124 +687,7 @@ export class EngramMcpServer {
...
-              type: "integer",
-              minimum: 1,
-              maximum: 200,
-              description: "Maximum results (default 50).",
-            },
-          },
-          additionalProperties: false,
-        },
-      },
+      ...WEARABLES_MCP_TOOLS, ...MEETINGS_MCP_TOOLS,
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/access-mcp.ts:690]: Acknowledged — captured as a follow-up per the maintainer hard-stop (lower-severity; not P0/P1 data-loss/isolation or a test failure).

File: packages/remnic-core/src/access-service.ts [hunk 20]
```diff
@@ -5729,60 +5743,62 @@ export class EngramAccessService {
...
+      .sync({ source: request.source, date: request.date, days: request.days, forceMemories: request.forceMemories });
+    // Manual one-shot sync (HTTP/MCP/CLI): the meeting tail-step arms a
+    // debounced, unref'd timer a short-lived caller exits before firing. Drain
+    // it now so a manual sync's meeting build actually runs before we return.
+    // The long-lived auto-sync daemon does NOT flush — it keeps coalescing.
+    if (this.orchestrator.config.meetings.enabled) {
+      await (await this.orchestrator.getMeetingsService(namespace)).flushBuilds();
+    }
+    return summaries;
   }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/access-service.ts:5776]: Fixed in 3e54c84a3 (#2123): meetingsBuild now enforces the READ ACL on the resolved namespace (via the existing resolveReadableNamespace resolver) in ADDITION to writableNamespaceFor, since the build reads the namespace's stored wearable transcripts + existing meeting records. A principal that may write a namespace but is omitted from its readPrincipals is now DENIED meetingsBuild with the shared EngramAccessInputError ('namespace is not readable'). resolveReadableNamespace is a no-op ACL when namespaces are disabled, so default/operator behavior is unchanged. Regression: meetings/namespace-isolation.test.ts 'a write-only principal (in writePrincipals, not readPrincipals) is DENIED meetingsBuild, while read+write and the machine owner succeed' (proven failing pre-fix: stashing the ACL line -> 'Missing expected rejection', passing post-fix).

File: packages/remnic-core/src/meetings/mcp-tools.ts [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+      "List stored meeting records (issue #1900), across all days or one day. Returns per-day meeting summaries; empty when meetings.enabled is off.",
+    inputSchema: {
+      type: "object",
+      properties: {
+        namespace: { type: "string", description: "Caller namespace to scope this operation to (issue #2123)." },
+        sessionKey: { type: "string", description: "Session key for principal/namespace resolution." },
+        date: {
+          type: "string",
+          description: "Optional day to list (YYYY-MM-DD). Omit for all days.",
+        },
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/meetings/mcp-tools.ts:24]: Acknowledged — captured as a follow-up per the maintainer hard-stop (lower-severity; not P0/P1 data-loss/isolation or a test failure).

File: packages/remnic-core/src/orchestration/workspace-ops.ts [hunk 22]
```diff
@@ -683,9 +718,138 @@ export class WorkspaceOpsCoordinator {
...
+            `meetings: tail-step build for ${date} failed: ${
+              err instanceof Error ? err.message : String(err)
+            }`,
+          ),
       });
+      this.deps.meetingsServiceByNamespace.set(namespace, service);
     }
-    return this.deps.wearablesServiceInstance;
+    return service;
   }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/workspace-ops.ts:853]: Acknowledged — captured as a follow-up per the maintainer hard-stop (lower-severity; not P0/P1 data-loss/isolation or a test failure).

File: packages/remnic-core/src/orchestrator.ts [hunk 23]
```diff
@@ -875,7 +876,8 @@ export class Orchestrator {
   private readonly consolidationObservers = new Set<
     (observation: ConsolidationObservation) => Promise<void> | void
   >();
-  private wearablesServiceInstance: WearablesService | null = null;
+  private readonly wearablesServiceByNamespace = new Map<string, WearablesService>();
+  private readonly meetingsServiceByNamespace = new Map<string, MeetingsService>();
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestrator.ts:880]: Acknowledged (Major-rated stability, Heavy-lift). Per maintainer hard-stop (code only for P0/P1 data-loss/isolation or a failing test), captured as a follow-up: getMeetingsService/getWearablesService register the namespace-keyed service only AFTER an await, so a first-call construction that completes AFTER destroy() can leave a debounced scheduler running; memoize the in-flight promise (or register the ns key before awaiting) so destroy() disposes in-flight construction. Narrow teardown race, not data-loss/isolation; no test failure.

File: packages/remnic-core/src/orchestrator.ts [hunk 24]
```diff
@@ -2996,10 +2998,9 @@ export class Orchestrator {
     return this.config.defaultNamespace;
   }
 
-  getWearablesService(): WearablesService {
-    return this.workspaceOpsCoordinator.getWearablesService(
-    );
-  }
+  getWearablesService(namespace: string = this.config.defaultNamespace): WearablesService { return this.workspaceOpsCoordinator.getWearablesService(namespace); }
+
+  getMeetingsService(namespace: string = this.config.defaultNamespace): Promise<MeetingsService> { return this.workspaceOpsCoordinator.getMeetingsService(namespace); }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestrator.ts:3003]: Acknowledged (Minor). Per maintainer hard-stop, captured as a follow-up: getWearablesService/getMeetingsService use a default-parameter fallback (applies only to undefined), unlike getStorage() which also maps empty-string namespace to config.defaultNamespace; normalize "" the same way so it doesn't create a separate empty-key cache entry. Edge-case only; no isolation/loss or test failure.

File: packages/remnic-core/src/meetings/memory-gen.ts [hunk 25]
```diff
@@ -0,0 +1,549 @@
...
+      for (const record of built) {
+        if (!unchanged.has(record.id)) {
+          toGenerate.push(record);
+          continue;
+        }
+        const episodePresent = await writer.hasMemoryFromSource(
+          meetingSourceLabel(record.id),
+          composeMeetingEpisodeContent(record),
+        );
+        if (!episodePresent) toGenerate.push(record);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/meetings/memory-gen.ts:531]: Acknowledged (P2). Per maintainer hard-stop (code only for P0/P1 data-loss/isolation or a failing test), captured as a follow-up: after a partial memory-write failure during fact generation, surface/re-queue so a later build retries rather than treating the source as complete. Best-effort today; the deterministic episode remains the presence sentinel for the existing retry path. No data-isolation/loss or test failure.

File: packages/remnic-core/src/orchestration/workspace-ops.ts [hunk 26]
```diff
@@ -635,12 +654,12 @@ export class WorkspaceOpsCoordinator {
...
-    if (!this.deps.wearablesServiceInstance) {
-      this.deps.wearablesServiceInstance = new WearablesService({
+  getWearablesService(namespace: string): WearablesService {
+    let service = this.deps.wearablesServiceByNamespace.get(namespace);
+    if (!service) {
+      service = new WearablesService({
         config: this.deps.config.wearables,
-        getStorage: async () =>
-          await this.deps.getStorageForNamespace(this.deps.bulkImportWriteNamespace()),
+        getStorage: async () => await this.deps.getStorageForNamespace(namespace),
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/workspace-ops.ts:662]: Acknowledged (P2, namespace reindex completeness). Per maintainer hard-stop, captured as a follow-up: after a caller-namespace wearable/meeting write, refresh the namespace-scoped search backend (not only the default) so caller-ns content is searchable immediately. Records ARE written correctly to the caller ns (no data-loss/isolation) — search-freshness only. No test failure.
